### ISSUE TITLE
The Draft Type-system is Too Anemic.

### ISSUE THREAD

**Issue by OneWingedShark** [microsoft/xlang#226]
I was reading the draft for the type-system ([XDN03](https://github.com/Microsoft/xlang/blob/master/design_notes/XDN03%20-%20xlang%20Type%20System.md)) and it looks simultaneously too much and too little — allow me to suggest that, as presented, the type-system is anemic in what might be expressed while having far too many fundamental types.

The goal of the type-system should not be about presenting familiar fundamental types (eg `Int16`) but rather giving an expressive type-system especially because the goal of xlang is about making cross-language APIs available to other languages and a good type-system may preclude errors. As an example, allow me to present a small section of Ada code:
```
   -- A Social Security Number type for use in a database; format: ###-##-####.
   Type Social_Security_Number is new String(1..11)
     with Dynamic_Predicate =>
       (For all Index in Social_Security_Number'Range =>
          (case Index is
             when 4 | 7  => Social_Security_Number(Index) = '-',
             when others => Social_Security_Number(Index) in '0'..'9'
          )
       );
```

The above declares a string-type which exactly conforms to the format of a social security number and thus does not require constant checking within the body of a function, not only does this formulation reduce the need for run-time checking and code-duplication, it also can be used to endure correctness of stored data [eg in the database].

Moreover, a more robust type-system could reduce the number of fundamental types needed; consider the effect the ability of parameterizing types can have — which has some overlap with #216 — the following psudeocode shows how all the signed/unsigned integers in the draft's fundamental-types could be reduced:
```
Type Generic_Signed_Integer(Bits : Positive) is range -2**(Bits-1)..2**(Bits-1)-1
     with Size => Bits;
Type Generic_Unsigned_Integer(Bits : Positive) is range 0..2**Bits-1
     with Size => Bits;
```
then the above could be used in a package/module containing
```
   Type Int8   is new Generic_Signed_Integer( 8);
   Type Int16  is new Generic_Signed_Integer(16);
   -- …
   Type UInt8  is new Generic_Unsigned_Integer( 8);
   Type UInt16 is new Generic_Unsigned_Integer(16);
   -- …
```
Not only would such paramaterization be extraordinarily useful for constructing generic types, reducing the need for more code [and introducing more chances for cut-and-paste style errors], but the facility could be used to increase safety as well. The following Ada construct, for example, would have completely precluded the possibility of the Heartbleed bug:
```
   -- The Message type takes the Length parameter and binds its value to the
   -- String's length, precluding the length/data mismatch; the default of
   -- filling the Text field's value with the space-character precludes the
   -- leakage that would be caused by using uninitialized memory.
   Type Message(Length : Natural) is record        -- Length 0 is an empty string.
      Text : String(1..Length) := (Others => ' '); -- Space-filled default.
   end record;
```

The fundamental String type in the proposal is, in a word, **bad**. Restricting strings to UTF16 is shortsighted and severely cripples the usefulness of the entire project, precluding libraries which use 8-bit character-sets or the usage of the full UTF32; if you're willing to allow types themselves as parameters to other types rather than mere values, you could say something like:
```
 Type Generic_String( Character : Type ) is Array(Positive range <>) of Character;
 Type UTF16 is new Generic_String( Chr_16 ); -- Assuming Chr_16 is a UTF16 codepoint.
```

Perhaps a look at the Interface Design Language by Wulf, Lamb, and Nestor (all of CMU) would prove instructive; there's a technical report [here](https://s3-eu-west-1.amazonaws.com/pstorage-cmu-348901238291901/12096773/file.pdf) and a book containing the specification [here](https://www.abebooks.com/servlet/BookDetailsPL?bi=22558478682&searchurl=isbn%3D0-7167-8198-0%26sortby%3D17&cm_sp=snippet-_-srp1-_-title1) (ISBN: 0716781980), I would highly recommend looking at them if for nothing else than the inspiration and insight into something similar, though tangential, to xlang's purpose. — If I were to build on IDL, personally, I would update its syntax to be similar to Ada/SPARK and leverage the SPARK proof-tooling.

Lastly, allow me to suggest xlang completely steal Ada's notion of "**subtype**" — in Ada [a type is a set of values conjoined with a set of operations](http://archive.adaic.com/standards/83rat/html/ratl-04-02.html#4.2), a subtype is a set (possibly null) of constraints upon the values of the [sub]type from which it is derived [which is equivalent to making a subset of the given values], thus you can say:
```
   -- A 32-bit integer.
   Type Integer is range -2**31..2**31-1 with Size => 32;
   -- A non-negative 32-bit integer.
   Subtype Natural is Integer range 0..Intger'Last;
   -- A positive 32-bit integer.
   Subtype Positive is Natural range Natural'Succ(Natural'First)..Natural'Last;

```
Perhaps some explanations of the above are needed. In Ada types have [attributes](https://en.wikibooks.org/wiki/Ada_Programming/Attributes) (see also [LRM — 4.1.4 Attributes](http://www.ada-auth.org/standards/12aarm/html/AA-4-1-4.html)), which can be queried to retrieve properties of the type. In the above the attribute `First` corresponds to the first value in the type, and likewise the `Last` attribute corresponds to the last element; the `Succ(X)` attribute returns the next element (after X) in the set of values and is thus equivalent to `X+1` for integer types, thus the above are saying "Natural is an Integer within [0..`<last integer value>`]" and "Positive is a Natural within [1..`<last natural value>`]", respectively. — This addition of constraints is incredibly useful for both reasoning as well as ensuring correctness, including this functionality in the xlang type-system would be very good for increasing the usability, especially with respect to such technologies as theorem-proovers.

**Comment by BrentRe** [comment]
Note that we are not trying to define an entirely new all-encompassing type system. Rather we are defining a type system that can be expressed in a natural and familiar way in various programming languages.

One of our conceptual tests prior to adding a new construct to the type system is how that construct could be expressed in various languages, such a JavaScript, C++, C#, Go, Rust, etc. 

The problem with leaning towards a more expressive type system is that it starts expressing type system constructs / constraints that are lost when projecting that type system to a language without the necessary semantics. It's always a tradeoff. We want to avoid a least-common denominator approach but we also don't want to introduce concepts which will be discarded by many languages.

As a simple example, both WinRT and xlang have the type system concept of a 64-bit integer, both signed and unsigned. However, those semantics are unable to be preserved when projecting the type to JavaScript, which has only a Number type with 56-bit precision. This leads to interoperability issues.

**Comment by contextfree** [comment]
@brentre One increasingly common expressive type system feature that I think you should consider is non-nullability constraints on reference types. This is becoming more and more common in both new and existing languages, for example TypeScript, Rust, Swift, and [upcoming additions to C#.](url=https://docs.microsoft.com/en-us/dotnet/csharp/nullable-references) . When a language expects non-nullability by default, it can be awkward to interact with API surfaces that may produce null values everywhere. 

I do think that to remain relevant and useful the WinRT/xlang type system should try to evolve along with trends among the languages that it works with, and I think non-nullability is an example of such a trend.

**Comment by raffaeler** [comment]
@BrentRe I am wondering about the opportunity for you to define the behavior of the cast/conversions among typed and untyped languages.
IMO it would be very useful (for everyone contributing to xlang) to provide a reccomended behavior for those transformations.
For example, when converting from the javascript number to C++, what are the guidelines in picking uint32_t, float or double? Internally nodejs/v8 makes a distinction between integers and floating point (for perf reasons). I believe xlang specs should give at least the guidelines to pick the best possible type.

**Comment by OneWingedShark** [comment]
**@BrentRe Said:**  
> The problem with leaning towards a more expressive type system is that it starts expressing type system constructs / constraints that are lost when projecting that type system to a language without the necessary semantics. It's always a tradeoff. We want to avoid a least-common denominator approach but we also don't want to introduce concepts which will be discarded by many languages.

Honestly, this seems completely backwards, akin to saying "*we need to optimize this language for compiler=writers, rather than for the programmers using it*" or "*we need to optimize for the lowest common denominator: Intel's 286*" — this stance/design-decision ***will*** kneecap the xlang project and render it nearly useless: after all, everybody else is going that route.

> As a simple example, both WinRT and xlang have the type system concept of a 64-bit integer, both signed and unsigned. However, those semantics are unable to be preserved when projecting the type to JavaScript, which has only a Number type with 56-bit precision. This leads to interoperability issues.

The solution is simple: make the language robust, make the projector raise an exception when it cannot translate into the target/projected language — the stated purpose of xlang is for APIs, are you going to make preconditions, postconditions, and function-ordering ("A must be called before B") simply because this concept doesn't exist in many languages? Yeah, it could be a very hard problem for projecting to some languages, but there's no reason you couldn't stub that particular functionality on those projectors or raise an exception or even put in the documentation: "*this advanced feature does not work in C, C++, [etc]*".

**@contextfree said:**  
> I do think that to remain relevant and useful the WinRT/xlang type system should try to evolve along with trends among the languages that it works with, and I think non-nullability is an example of such a trend.

I agree, but with a caveat: xlang ought to be *very* high-level, able to express complicated things, even if they aren't directly translatable into the lowest common denominator language. BrentRe's reply honestly reminds me of the stupidity that is the design of WebAssembly: they're literally designing it for C++, since that's their minimal viable product, despite that this kneecaps them with regard to parallelism and a **lot** of optimization. (See: Guy Steele's [How to Think about Parallel Programming: Not!](https://www.infoq.com/presentations/Thinking-Parallel-Programming) video.)

**@raffaeler said:**  
> I am wondering about the opportunity for you to define the behavior of the cast/conversions among typed and untyped languages.

This is an excellent question, and while I'm not sure cast/conversions are a property of the type-system *per se* it is the case that defining them well can alleviate a lot of confusion, especially if you're operating in a multilanguage environment with languages that have very different conversion rules.

> For example, when converting from the javascript number to C++, what are the guidelines in picking uint32_t, float or double? Internally nodejs/v8 makes a distinction between integers and floating point (for perf reasons). I believe xlang specs should give at least the guidelines to pick the best possible type.

This should be handles automatically by the projector... but in this particular case JavaScript numerics are mandated by [ECMA-262](https://www.ecma-international.org/publications/files/ECMA-ST/Ecma-262.pdf) to be:

> **4.3.20 Number value**  
> primitive value corresponding to a double-precision 64-bit binary format IEEE 754-2008 value
> 
> 
> **4.3.21 Number type**  
> set of all possible Number values including the special “Not-a-Number” (NaN) value, positive infinity, and negative infinity

So, the answer is `double`, IIRC.

This illustrates my point: it would be stupid to force every projector to use the binary 64-bit IEEE-754 for a dice game because of the Javascript restriction despite *most* languages having a proper integer type, and some languages having ranged types like `Type D6 is range 1..6;`.

xlang should think in terms of the problem-space (that is representing APIs), giving tools and constructs that are robust and expressive, rather than in terms of what common/popular languages can do.

**Comment by raffaeler** [comment]
@OneWingedShark the V8 javascript execution engine internally distinguish between integers and floats even if javascript claims to only use IEEE-754 for all.
The distinction is super-important for perf reasons: using a double instead of int in a for loop would lead to very poor performance.
The conclusion is that javascript implicitly uses two different numeric types even if there is no explicit way to define them. While it is desiderable leaving the deduction to V8, as a xlang provider developer, I definitely want to get advantage from this and provide a mapping to C# not just using doubles but, at least, integers and doubles.
My request is to define a best practice to guide xlang language providers to adopt a reasonable choice when mapping types that **apparently** cannot be mapped.
I hope to have clarified a bit more my point.

**Comment by BenJKuhn** [comment]
@OneWingedShark, thanks for raising the discussion.

I don't think your suggestion is at odds with what we're trying to accomplish, but we're not there yet either. To achieve interoperability and fine grained control, we're still focused more on the ABI and basic language interoperability. They're required building blocks. When you transfer data from one type system to another you need to agree on the format and interpretation of the data. The IDL-style tooling we're building now will also be focused on that for the time being. 

However, the metadata format and abstractions will allow us a lot of room to grow. Trying to achieve that level of type system expression in the languages that we're currently targeting would greatly slow our ability to get to V1 (and given that Windows ships thousands of APIs using a very type system, it's obviously sufficient to provide a lot of value), but I think the door is still open to achieve something along the lines of what you're proposing this in future iterations, and I can certainly see the value. 

It could also lead to a great set of topics around enforcement of debuggability of pre-conding, post-condition and invariant violations, which is something that WinRT is not highly consistent about (e.g. handling of non-null values). I think there's room for some innovation here in both the type system & the runtime.

Ben

**Comment by OneWingedShark** [comment]
> @OneWingedShark, thanks for raising the discussion.

You're quite welcome.

> It could also lead to a great set of topics around enforcement of debuggability of pre-conding, post-condition and invariant violations, which is something that WinRT is not highly consistent about (e.g. handling of non-null values). I think there's room for some innovation here in both the type system & the runtime.

I very much agree, and I see a lot of potential for the underlying idea.

> However, the metadata format and abstractions will allow us a lot of room to grow. Trying to achieve that level of type system expression in the languages that we're currently targeting would greatly slow our ability to get to V1 (and given that Windows ships thousands of APIs using a very type system, it's obviously sufficient to provide a lot of value), but I think the door is still open to achieve something along the lines of what you're proposing this in future iterations, and I can certainly see the value.

This is where it really hurts: these things aren't all that good as a "bolt-on". Take, for example, security: you _can_ do a program slap-dash then come back later and do bug-fixes, refactorings, and re-engineering to patch-up all the security-holes in the initial implementation **OR**  you could design the program to be secure from the outset. The latter is _much_ more likely to cost less, in the long run, but the former is deceptively alluring because you get "working" easier/quicker.

Or, if you'll allow an anecdote:
Several years back I was talking with someone on an online forum, and we were talking about tech and Ada came up. He related to me that he was part of a company that did a code-audit on Windows (2.x, IIRC; 98% sure it was pre _Windows for Workgroups 3.11_) he and his company noted a lot of possible errors and recommended that it be rewritten in Ada. — Now, if Microsoft **_had_** done a rewrite in Ada, there would have been several major consequences: (1) the massive amount of buffer-overflows that were notorious for awhile wouldn't have happened; (2) the OS would have been more stable, esp. if exceptions were incorporated into the process execution-type; (3) the Windows message-loop could have been vastly improved [though to be fair, the `task interface` that would have been best for this wasn't introduced until [Ada 2005](https://www.adaic.org/resources/add_content/standards/05rm/html/RM-3-9-4.html), and [Smalltalk's message-passing](https://stackoverflow.com/q/42498438/608963) is arguably even better, though might have been too heavy-weight/dynamic/slow at that time-period]; and (4) the recent "_everything has to be rewritten for multicore!!_" craze would have been trivially solvable by recompiling the source thanks to the high-level `task` construct.

Now I understand that there's thousands of APIs that Windows ships, this is inevitable when you have as large a code-base that's been used for so long, but how much value-added would it be if you _could_ trivially say "parameter `X` cannot be null" in the source and have the projectors take care of this? — How valuable would it be to be able to, for a binding of OpenGL, be able to [specify for `glAccum`](https://www.khronos.org/registry/OpenGL-Refpages/gl2.1/xhtml/glAccum.xml) the following?

    Type Accumulator_Operators is (
		ACCUM => GL_ACCUM,	LOAD => GL_LOAD,	OP_RETURN => GL_RETURN,
		MULT => GL_MULT,		ADD => GL_ADD
        ) with Size => GL_Enum'Size; -- Where GL_* are the appropriate constants.
    --...
    Procedure Accumulate( Operation : Accumulator_Operators; Value : GL_Float )
      with Import, Convention => stdcall, External_Name => "glAccum",
      Pre => not Executing; -- Executing is true between calls of glBegin & glEnd.

The above eliminates in client-code `GL_INVALID_ENUM` and `GL_INVALID_OPERATION` errors, leaving only the `GL_INVALID_OPERATION` as a possible error. — So, just by specifying things at a higher level we can eliminate several error-classes altogether, making things both safer and more maintainable than the low-level specification.

> I don't think your suggestion is at odds with what we're trying to accomplish, but we're not there yet either. To achieve interoperability and fine grained control, we're still focused more on the ABI and basic language interoperability. They're required building blocks. When you transfer data from one type system to another you need to agree on the format and interpretation of the data. The IDL-style tooling we're building now will also be focused on that for the time being.

**Is** the ABI that big of a concern?
_Why_ is ABI that big of a concern?

You're asserting they're fundamental, basic to interoperability... but is it?
Isn't it true that changing compilers can destroy the binary-compatibility? Isn't it true that changing architectures can destroy the binary compatibility? (eg moving from a 32-bit compiler to a 64-bit compiler, or from x86_64 to IA64.) — Given that, it seems rather like a premature optimization to worry about ABI, and might I suggest that the problem might be addressed as parameters of the architecture fed into the projector and/or properties of the IDL-like source? What would the following be missing in terms of information needed for interoperability?

    Function Pop( Object : in out Stack ) return Element
      with Convention => COBOL, External_Name => "POP-ELEMENT",
           Pre => Not Empty(Object),
           Post => Pop'Result = Element(Object'Old) and Size(Object) = Size(Object)-1;

And what about "prior art" in the field? IIRC, IBM's _[System Object Model](https://infogalactic.com/info/IBM_System_Object_Model)_ was extremely good at binary-compatibility; perhaps it would be easier to obtain a license from them for that technology? Or, perhaps, OpenVMS's _[Common Language Environment](https://www.itec.suny.edu/scsys/vms/OVMSDOC0731/ovms_archived/OVMS_PROG_ENVIRON.PDF)_, which IIUC is now under the care of [VMS Software, Inc.](https://vmssoftware.com/contact.html)

**Comment by BenJKuhn** [comment]
Oh wow, I see CORBA's mentioned in there too. I worked on an early attempt at a CORBA / SOAP bridge product as an intern nearly 20 years ago at a previous company. That brings back memories. There's a great book comparing other similar systems called Component Software written by Clemens Szyperski. Although a bit dated now, it's still a good overview of the problem domain.

You're hitting on a key fundamental question about the nature of this project though. xlang is, at the end of the day, a component system that happens to support many languages rather than a language interop model fist. That leads you down a path similar to COM, SOM, CORBA, etc. A comparison between SOM & xlang would be interesting. WinRT is a bit more modern of a foundation to start from in many respects, but there may be things to learn from what worked & didn't in SOM, since it starts from a very different foundation than the COM lineage. Am I correct in otherwise observing that SOM seems to be abandonware (last release in the 90s?). An example of an interop-first approach is [swig](http://swig.org/). It's a robust language interop mechanism between C and pretty much everything else, but certainly does not get you a component model. They make some very different design decisions. 

Binary compatibility requires some amount of common context for the compiled binaries: typically the same processor architecture and operating system. With that framing in mind, it's quite achievable, as has been proven out over time on Windows. COM has been the extensibility mechanism for many applications & services on Windows. WinRT was a great Windows-specific refresh to COM that allows a much simpler developer experience. xlang is, in some sense, the cross-plat "refresh" of WinRT. Being platform agnostic, portable, and open-source appeals to a much larger potential audience and embraces new use cases, but it also very much follows in that lineage.

I'm pretty sure if you tried you could re-write the expression above in a C / C# like syntax that could be comfortable accommodated in IDL without feeling like a bolt-on. [Other languages ](https://www.eiffel.org/doc/eiffelstudio/Precondition) have done a reasonable job with even uglier syntax conventions.

Ben

**Comment by OneWingedShark** [comment]
> Oh wow, I see CORBA's mentioned in there too. I worked on an early attempt at a CORBA / SOAP bridge product as an intern nearly 20 years ago at a previous company. That brings back memories.

Good memories?

> There's a great book comparing other similar systems called Component Software written by Clemens Szyperski. Although a bit dated now, it's still a good overview of the problem domain.

Just because it's old doesn't mean that its wrong or useless though. To get on a slight tangent to prove the point, the idea of DB-backed development-environments completely solves both the diff-tool false-positive and Continuous Integration problems (and better than "modern" approaches), as detailed in [this paper from 1987](http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.26.2533&rep=rep1&type=pdf). — But thank you for pointing out this book, it looks interesting.

> You're hitting on a key fundamental question about the nature of this project though. xlang is a component system that happens to support many languages rather than a language interop model fist.

I think I see the distinction, but would like a bit more elaboration on it.
Also, is this a valid distinction? Or, rather, should I phrase it as "**_Do the two necessarily have to be opposed?_**"

That components are needed is undisputed, but the concentration on implementation details first (eg ABI and lowest-common denominator type-system) seems backwards to me: just like if it were interop-first the primary focus should be not describing thing in [primarily] a low-level form, but in the ability to describe a wide range of things [ie high-level] foremost. — So, for a component-based system the focus should be primarily on the ability of describing components and their interactions at a high-level rather than prematurely saying (eg) "_enum is 16 bits_."

If you remember Delphi, there was a DFM-file which contained properties of the [visual] components which were set to non-default values — I don't remember if C# has the same sort of setup, but given Anders Hejlsberg's involvement in its design it's rather likely — and that could be considered the text-serialization of the component. If you were to take this idea and marry it to the base meta-object of the component framework, integrating/adding a `serialize`/`deserialize` pair of functions which themselves use/implement ASN.1… well, now you have a platform-independent way to load, store, distribute, and transfer components, regardless of the [eventual] ABI —— Heck, if you were to split the triggered 'actions' [eg `OnClick`] into three parts (1) client/pre-process, (2) server/process, & (3) client/post-process you could have the framework very naturally do client-server distributed applications; the old (and now new again) idea of having a lightweight device do the UI and light processing with all the "hard work" offloaded to your big computer.

So, it seems to me that you could do a good component-based system completely independently of things like ABI.

> That [interop-first model] leads you down a path similar to COM, SOM, CORBA, etc. A comparison between SOM & xlang would be interesting.

I fully agree, it would be _quite_ interesting; I just don't think the two are necessarily opposed — if you did the above, but also used (eg) [VMS's CLR](https://infogalactic.com/info/OpenVMS#Common_Language_Environment) [or COM or SOM or whatever] to solve the ABI problem, wouldn't that essentially be addressing the whole thing?

> WinRT is a bit more modern of a foundation to start from in many respects, but there may be things to learn from what worked & didn't in SOM, since it starts from a very different foundation than the COM lineage.

This is true, but "_modern_" doesn't mean a whole lot when it comes to quality of design — as an example, consider the virtual keyboard that Apple-products use: note how the J and F keys are rendered with the little touch-ridge? That ridge is on physical keyboards so that you can feel where to place your fingers, meaning that it's an absolute waste on the flat/featureless touchscreen of an ipad… and, likewise, a lot of "_modern_" solutions are strictly inferior to older solutions.

Perhaps a good illustration of this can be had with JSON. JSON is a terrible solution to the problem of serialization/deserialization, despite its modernity and popularity: it lacks the ability to constrain values, it lacks the ability to [automatically] validate the data received, it doesn't _really_ support JavaScript objects (because numerics cannot be NaN or INF, but must [possibly] be in JavaScript.)

TL;DR — "Modern" is **not** a good indicator of quality, but the comparison between (eg) SOM & xlang would be quite interesting.

> Am I correct in otherwise observing that SOM seems to be abandonware (last release in the 90s?).

It looks like it; in the [admittedly small] efforts to track it down on my part I haven't come across new/updated info… but my search-abilities aren't the greatest and IBM could certainly "pocket" or "rebrand" the tech on the downlow w/o anyone knowing.

> An example of an interop-first approach is [swig](http://swig.org/). It's a robust language interop mechanism between C and pretty much everything else, but certainly does not get you a component model. They make some very different design decisions.

Interesting.

> Binary compatibility requires some amount of common context for the compiled binaries: typically the same processor architecture and operating system.

Sure; but with a little clever design, the engineering can mitigate this. (As shown above WRT adding cross-platform serialization via ASN.1.)

> With that framing in mind, it's quite achievable, as has been proven out over time on Windows. COM has been the extensibility mechanism for many applications & services on Windows.

Granted.
Microsoft has done an amazing job WRT the backwards-compatibility of Windows. (Arguably too good, as so many of its programs are essentially forced to the x86[_64] architecture because of their historical reasons.)

> WinRT was a great Windows-specific refresh to COM that allows a much simpler developer experience. xlang is, in some sense, the cross-plat "refresh" of WinRT. Being platform agnostic, portable, and open-source appeals to a much larger potential audience and embraces new use cases, but it also very much follows in that lineage.

I can see this. I still think that basing the type-system on lowest-common denominator features is a self-defeating move though.

> I'm pretty sure if you tried you could re-write the expression above in a C / C# like syntax that could be comfortable accommodated in IDL without feeling like a bolt-on. [Other languages ](https://www.eiffel.org/doc/eiffelstudio/Precondition) have done a reasonable job with even uglier syntax conventions.

What, exactly, does syntax have to do with anything here?
Forget, for a moment, "C/C# like" (or IDL) and consider the conceptual-space — it's about being able to specify something first and foremost, not about what the method-of-specifying looks like, right?

Secondly, what IDL? [There](https://infogalactic.com/info/IDL_specification_language) [are](https://infogalactic.com/info/OMG_IDL) [several](https://infogalactic.com/info/Franca_IDL). — More seriously, anyone on this project should take a look at the Wulf/Lamb/Nestor IDL ([tech-report](https://s3-eu-west-1.amazonaws.com/pstorage-cmu-348901238291901/12096773/file.pdf), [book](https://www.barnesandnoble.com/w/automating-interfaces-in-a-software-system-richard-snodgrass/1000783797)), which was remarkably expressive for being developed [circa 1981](https://apps.dtic.mil/dtic/tr/fulltext/u2/a228821.pdf) —link to DTIC.mil-archived development plan for DIANA (Descriptive Intermediate Attributed Notation for Ada; essentially an Attribute-grammar+IR hybrid) & [STARS](https://apps.dtic.mil/docs/citations/ADB157222) (Software Technology for Adaptable, Reliable Systems).
